lib/c-api/src/wasm_c_api/externals/function.rs

```diff
@@ -124,21 +138,29 @@ pub unsafe extern "C" fn wasm_func_delete(_func: Option<Box<wasm_func_t>>) {}
 #[no_mangle]
 pub unsafe extern "C" fn wasm_func_call(
     func: &wasm_func_t,
-    args: *const wasm_val_t,
-    results: *mut wasm_val_t,
+    args: &wasm_val_vec_t,
+    results: &mut wasm_val_vec_t,
 ) -> Option<Box<wasm_trap_t>> {
-    let num_params = func.inner.ty().params().len();
-    let params: Vec<Val> = (0..num_params)
-        .map(|i| (&(*args.add(i))).try_into())
-        .collect::<Result<_, _>>()
-        .ok()?;
+    let params = args
+        .into_slice()
+        .map(|slice| {
+            slice
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<Vec<Val>, _>>()
+                .expect("Argument conversion failed")
```

Comment:
We used to report these failures as errors to the user: why is it `TryInto` instead of `Into`? Is this failure something users could encounter?

Reply:
Yes, because `wasm_val_t` contains an enum that we must build from `u8`. We can out-of-bounds values. A `TryInto` is appropriate here.